extension/js/content_scripts/webmail/gmail-element-replacer.ts

```diff
@@ -520,7 +546,107 @@ export class GmailElementReplacer implements WebmailElementReplacer {
+        if (this.debug) {
+          console.debug('processAttachments() try -> awaiting chunk + awaiting type');
+        }
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        const data = await this.queueAttachmentChunkDownload(a).result;
+        const openpgpType = MsgUtil.type({ data });
+        if (openpgpType && openpgpType.type === 'publicKey' && openpgpType.armored) {
+          // if it looks like OpenPGP public key
+          treatAs = 'publicKey';
+        } else if (openpgpType && ['encryptedMsg', 'signedMsg'].includes(openpgpType.type)) {
+          treatAs = 'encryptedMsg'; // todo: signedMsg ?
+        } else {
+          attachmentSel.show().children('.attachment_loader').text('Unknown OpenPGP format');
+          if (this.debug) {
+            console.debug("processAttachments() try -> awaiting done and processed -- doesn't look like OpenPGP");
+          }
+          return 'shown';
+        }
+        if (this.debug) {
+          console.debug('processAttachments() try -> awaiting done and processed');
+        }
+      }
+      if (treatAs !== 'plainFile') {
+        this.hideAttachment(attachmentSel, attachmentsContainerInner);
+      }
+      if (treatAs === 'hidden') {
+        return 'hidden';
+      } else if (treatAs === 'encryptedFile') {
+        // actual encrypted attachment - show it
+        attachmentsContainerInner.prepend(this.factory.embeddedAttachment(a, true)); // xss-safe-factory
+        return 'replaced'; // native should be hidden, custom should appear instead
+      } else if (treatAs === 'encryptedMsg') {
+        msgElReference.msgEl = /* xss-safe-factory */ this.updateMsgBodyEl_DANGEROUSLY(
+          msgElReference.msgEl,
+          'set',
+          this.factory.embeddedMsg('encryptedMsg', '', msgId, false, senderEmail)
+        );
+        return 'hidden'; // native attachment should be hidden, the "attachment" goes to the message container
+      } else if (treatAs === 'publicKey') {
+        // todo - pubkey should be fetched in pgp_pubkey.js
+        return await this.renderPublicKeyFromFile(a, attachmentsContainerInner, msgElReference.msgEl, isOutgoing, attachmentSel);
+      } else if (treatAs === 'privateKey') {
+        return await this.renderBackupFromFile(a, attachmentsContainerInner, msgElReference.msgEl);
+      } else if (treatAs === 'signature') {
+        // todo: generate frameId here, prevent frameId duplicates?
+        const { frameId, frameHtml: embeddedSignedMsgXssSafe } = this.factory.embeddedRenderMsg();
+        msgElReference.msgEl = this.updateMsgBodyEl_DANGEROUSLY(msgElReference.msgEl, 'set', embeddedSignedMsgXssSafe); // xss-safe-factory
+        const frameWindow = XssSafeFactory.getWindowOfEmbeddedMsg(frameId);
+        if (frameWindow) {
+          const renderModule = new RenderRelay(frameWindow);
+          this.processSignedMessage(msgId, renderModule, senderEmail).catch(Catch.reportErr); // todo: clear cached items
```

Comment:
Here is a demo of the `GmailElementReplacer`-context signature verification I suggested.
It is currently triggered by an attachment that returns `treatAs()==='signature'` -- not an optimal way probably, but ok for the demo.

The idea is to insert a lightweight `pgp_render_block` frame instead of `pgp_block` that will only do rendering/printing, while decryption/verification is handled in a intermediate class `MessageRenderer` (probably should be renamed to `MessageProcessor`) that can be called by either `GmailElementReplacer` or `InboxActiveThreadModule`, so the logic from `PgpBlockViewDecryptModule` and `PgpBlockViewSignatureModule` etc. should be moved to `MessageRenderer`. I have done some work in this direction, but that's not easy as these modules are tightly coupled among other `PgpBlockView*Module`s.

Rendering -- The decryption/verification procedure sends notification to the iframe by means of `postMessage`:
- need to review security concerns (the receiving iframe can display a "signed" message sent by some other script, how dangerous is that?)
- probably need to stringify JSON and base64-encode with regard to #2587, test the message size limits
- messages may be lost if sent too early (before the iframe finished loading) -- so there is a need to queue messages until the iframe has signaled it's ready tor receive messages.

Are we on the same page here, @tomholub?
I can try and do `PgpBlockView*Module` decoupling in a separate PR to reduce the size of this one, if needed.
 cc @sosnovsky

Reply:
> need to review security concerns (the receiving iframe can display a "signed" message sent by some other script, how dangerous is that?)

what about adding some token with initial request, which will be sent back along with `RenderMessage`? then receiving frame will be able to check if received message is the right one.

overall direction looks good for me, new implementation looks more straightforward than the current one with multiple `pgp_block`

> I can try and do PgpBlockView*Module decoupling in a separate PR to reduce the size of this one, if needed.

It'll be good if possible and won't affect functionality in this PR

Reply:
as long as by "other scripts" you mean other pages of our own extension, then that's fine. doesn't need to be addressed. 

of course if it was handling requests sent by software we didn't write (eg other extensions) that would be a big problem.

Reply:
for the rest, i'd roma says it's good, then i'm good

Reply:
> as long as by "other scripts" you mean other pages of our own extension, then that's fine. doesn't need to be addressed.
> 
> of course if it was handling requests sent by software we didn't write (eg other extensions) that would be a big problem.

The receiver can check `event.origin`, and it is something like `mail.google.com`, so other extensions could possibly "impersonate" our content script if only they send a message from their content script installed to `mail.google.com` as well.

Reply:
> what about adding some token with initial request,

I guess the URL (query string) with which the frame was created is visible to every script having access to DOM, no?

It may be possible to include a sort of "public key" in the frame query string, so the frame can check the "signature" of messages from the content script, I'm wondering what overhead would be and how to accomplish this since we don't want to load OpenPGP.js into each frame.

Reply:
let's not over complicate this

if today the browser extension actually responds to messages sent from order extensions, then that's a HUGE problem and needs to be addressed by somehow forbidding that, in manifest or by always checking the origin. regardless of this PR. 

if today that does not happen, then none of this is a problem and we can proceed without further checks